### PR TITLE
nixos: escape mount options

### PR DESCRIPTION
**Pull request by Atemu**
Some mount options might include path names and those often contain spaces and therefore must be escaped. An example which prompted me to make this change is the path of a btrfs subvolume.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Atemu** [comment]
Things are never that simple, are they...

This doesn't work in initrd for some reason.

**Comment by Atemu** [comment]
Problem found: stage-1 builds up its own fstab from the declared fsinfo file as it goes through it and then relies on `mount`'s fstab parsing.

This is odd but I'm going to leave it as is and escape those strings in fsinfo. I don't want to touch stage-1. systemd in initrd shouldn't have any of these issues since it uses the fstab generated in filesystems and that's the future anyways.

**Comment by Atemu** [comment]
That didn't end up working out. Bash doesn't like the backslash...

I've attempted to escape things within the script now.

**Comment by Atemu** [comment]
Yeah, that works.

**Comment by Atemu** [comment]
All boot tests except for ubootExtlinux pass but that appears to already be broken on my PR's base aswell.

Of course I tested this manually too.

**Comment by RaitoBezarius** [comment]
Can you provide a test that fails without this PR which is supposedly a test which pass in that PR?

**Comment by Atemu** [comment]
As explained in the commit message, a btrfs subvolume with a space in it could not be mounted before. An example of this would be `subvol=foo bar,noatime,discard=async".

If you build a NixOS generation with such options set for one of the `fileSystems`, you'd get an invalid fstab. This is simple to test without deploying it in a system: Just `nixos-rebuild build` and take a look at `result/etc/fstab`. Spaces delimit fields in fstab, so if the options have spaces in them, that's invalid. Spaces needs to be substituted with `\040`.

In stage-1 the same problem applied. Additionally, the mount point could also not have a space in it which I also fixed while I was at it.

**Comment by lilyinstarlight** [comment]
> As explained in the commit message, a btrfs subvolume with a space in it could not be mounted before. An example of this would be `subvol=foo bar,noatime,discard=async`.

Do you think you could write a NixOS test for this? I could probably provide a simple diff for one if you need. Otherwise I think this PR looks good

**Comment by Atemu** [comment]
That's a good idea, yeah. Is there a stage-1 or initrd test I could extend where checking this sort of thing makes sense? I've found a btrfs test that could use a mount-point with space in it but it doesn't test initrd.

Since that'd be my first NixOS test, I'd rather do that in a separate PR and get this one in before branch-off. Perhaps I'd also wait for the new test infra that will come after 22.11 in the soonest.

File: nixos/modules/system/boot/stage-1-init.sh
Comment (lilyinstarlight) [inline nixos/modules/system/boot/stage-1-init.sh:351]:
```suggestion
    # Replace space
    escaped="${original// /\\040}"
    # Replace tab
    echo "${escaped//$'\t'/\\011}"
```

You need to replace all matches here, not just the first. Otherwise this PR looks good and a test I wrote (diff below -- feel free to include it in this PR) passes for both old stage-1 and systemd stage-1 after this change

<details>
<summary>installer-subvol-space-test.patch</summary>

```diff
diff --git a/nixos/release-combined.nix b/nixos/release-combined.nix
index bd7b452735f..337b5192776 100644
--- a/nixos/release-combined.nix
+++ b/nixos/release-combined.nix
@@ -77,6 +77,7 @@ in rec {
         (onFullSupported "nixos.tests.i3wm")
         (onSystems ["x86_64-linux"] "nixos.tests.installer.btrfsSimple")
         (onSystems ["x86_64-linux"] "nixos.tests.installer.btrfsSubvolDefault")
+        (onSystems ["x86_64-linux"] "nixos.tests.installer.btrfsSubvolEscape")
         (onSystems ["x86_64-linux"] "nixos.tests.installer.btrfsSubvols")
         (onSystems ["x86_64-linux"] "nixos.tests.installer.luksroot")
         (onSystems ["x86_64-linux"] "nixos.tests.installer.lvm")
diff --git a/nixos/tests/installer-systemd-stage-1.nix b/nixos/tests/installer-systemd-stage-1.nix
index d02387ee80e..03f0ec8d746 100644
--- a/nixos/tests/installer-systemd-stage-1.nix
+++ b/nixos/tests/installer-systemd-stage-1.nix
@@ -8,9 +8,10 @@
   # them when fixed.
   inherit (import ./installer.nix { inherit system config pkgs; systemdStage1 = true; })
     # bcache
-    # btrfsSimple
-    # btrfsSubvolDefault
-    # btrfsSubvols
+    btrfsSimple
+    btrfsSubvolDefault
+    btrfsSubvolEscape
+    btrfsSubvols
     # encryptedFSWithKeyfile
     # grub1
     # luksroot
diff --git a/nixos/tests/installer.nix b/nixos/tests/installer.nix
index d9f64a781c5..9b3c8a76299 100644
--- a/nixos/tests/installer.nix
+++ b/nixos/tests/installer.nix
@@ -911,4 +911,25 @@ in {
       )
     '';
   };
+
+  # Test to see if we can deal with subvols that need to be escaped in fstab
+  btrfsSubvolEscape = makeInstallerTest "btrfsSubvolEscape" {
+    createPartitions = ''
+      machine.succeed(
+          "sgdisk -Z /dev/vda",
+          "sgdisk -n 1:0:+1M -n 2:0:+1G -N 3 -t 1:ef02 -t 2:8200 -t 3:8300 -c 3:root /dev/vda",
+          "mkswap /dev/vda2 -L swap",
+          "swapon -L swap",
+          "mkfs.btrfs -L root /dev/vda3",
+          "btrfs device scan",
+          "mount LABEL=root /mnt",
+          "btrfs subvol create '/mnt/nixos in space'",
+          "btrfs subvol create /mnt/boot",
+          "umount /mnt",
+          "mount -o 'defaults,subvol=nixos in space' LABEL=root /mnt",
+          "mkdir /mnt/boot",
+          "mount -o defaults,subvol=boot LABEL=root /mnt/boot",
+      )
+    '';
+  };
 }
```
</details>

**Review by lilyinstarlight — CHANGES_REQUESTED** [review]

**Comment by Atemu** [comment]
I'll squash and check out the test when I get to my Linux machine later today or tomorrow.

**Comment by Atemu** [comment]
This is ready to go I think.

**Review by lilyinstarlight — APPROVED** [review]
The tests all pass now (including the other btrfs ones that were previously still disabled for systemd-initrd for some reason) and this change looks good to me!

Since this does add a new test to the `tested` jobset (which I mostly did since the other btrfs installer tests were all in it), hopefully someone more authoritative can give it a quick look and okay it/merge it (I'm not a committer and cannot merge PRs)

**Review by ElvishJerricco — APPROVED** [review]
I don't really have an opinion on the scripted stage 1 implementation, but seeing the systemd stage 1 btrfs installer tests pass this new test looks good to me.

**Comment by Atemu** [comment]
Hi, could we merge this? Just got bitten by this again because I forgot to backport these changes to my fork.

File: nixos/modules/system/boot/stage-1-init.sh
Comment (Luflosi) [inline nixos/modules/system/boot/stage-1-init.sh:349]:
```suggestion
    local escaped="${original// /\\040}"
```

**Review by Luflosi — APPROVED** [review]
LGTM.

**Comment by Atemu** [comment]
I think at that point you must be intentionally trying to break the system ;)

I'm not actually sure you could escape that if you wanted to. The fstab format isn't exactly a flexibly one.

**Comment by xaverdh** [comment]
> I think at that point you must be intentionally trying to break the system ;)
> 
> I'm not actually sure you could escape that if you wanted to. The fstab format isn't exactly a flexibly one.

Well the state after this pr is arguably more correct than before, so this should not be blocked by such considerations. Still looking at `man fstab`, I am a bit surprised by the lack of consideration there..

**Comment by xaverdh** [comment]
For reference, [here](https://github.com/util-linux/util-linux/blob/441f9b9303d015f1777aec7168807d58feacca31/lib/mangle.c#L60) is the code in `util-linux` that parses fstab. It actually translates all `\[...]` sequences where `[..]` is three octal values.

**Comment by xaverdh** [comment]
> For reference, [here](https://github.com/util-linux/util-linux/blob/441f9b9303d015f1777aec7168807d58feacca31/lib/mangle.c#L60) is the code in `util-linux` that parses fstab. It actually translates all `\[...]` sequences where `[..]` is three octal values.

So I guess a correct way to do things, would be to in addition replace all occurrences of the literal `\`, by `\134` (the corresponding octal code in ascii).

**Comment by Atemu** [comment]
> I am a bit surprised by the lack of consideration there..

I'm not, fstab is *ancient*. It first appeared in 4.0BSD which is from [1980](https://gunkies.org/wiki/4BSD).

> For reference, [here](https://github.com/util-linux/util-linux/blob/441f9b9303d015f1777aec7168807d58feacca31/lib/mangle.c#L60) is the code in `util-linux` that parses fstab. It actually translates all `\[...]` sequences where `[..]` is three octal values.

We don't only have util-linux but also whatever minimal `mount` we have in initrd. Not sure whether that's from toybox, busybox or whatever but nothing that's linked against glibc I think.

I won't be pursuing escaping escape sequences or other extremely odd file names (newlines?) in this PR. I only care about file names you might actually find in a real system like spaces and mayybe tabs. We don't escape those in our runtime fstab either; I reused that escaping function.   
Feel free to do that in a follow-up though @xaverdh.

**Comment by xaverdh** [comment]
> > I am a bit surprised by the lack of consideration there..
> 
> I'm not, fstab is _ancient_. It first appeared in 4.0BSD AFAICT which is from [1980](https://gunkies.org/wiki/4BSD).
> 
> > For reference, [here](https://github.com/util-linux/util-linux/blob/441f9b9303d015f1777aec7168807d58feacca31/lib/mangle.c#L60) is the code in `util-linux` that parses fstab. It actually translates all `\[...]` sequences where `[..]` is three octal values.
> 
> We don't only have util-linux but also whatever minimal `mount` we have in initrd. Not sure whether that's from toybox, busybox or whatever but nothing that's linked against glibc I think.
> 
> I won't be pursuing escaping escape sequences or other extremely odd file names (newlines?) in this PR. I only care about file names you might actually find in a real system like spaces and mayybe tabs. We don't escape those in our runtime fstab either; I reused that escaping function. Feel free to do that in a follow-up though @xaverdh.

We use `busybox` in the [initrd](https://github.com/oxij/nixpkgs/blob/456f97f2e699511f4ff7a81b4e4b61cc66e62863/nixos/modules/system/boot/stage-1.nix#L104), which also allows escaping `\` (c.f. https://sourceware.org/git?p=glibc.git;a=blob;f=misc/mntent_r.c;h=39fa08f2c61cf57334c3602e0104509a519d7f96;hb=9798906a426fc458b949271bcc9b8ad1608de867#l97).
But I agree, that can be left to a follow up pr.

**Review by mweinelt — APPROVED** [review]

**Comment by Atemu** [comment]
Thank you!

**Comment by pupbrained** [comment]
I had to make a fork with this PR reverted due to any iterations made after it was merged failing to boot.

It might be a problem on my end, but I'm not sure. Without the PR everything works fine, but with it `/mnt-root/` is not created and therefore gives a "No such file or directory" error during stage 1, and a kernel panic related to the init process being killed if that error is ignored.

**Comment by lilyinstarlight** [comment]
@pupbrained, can you try separating these options out into multiple strings rather than as arrays with one singular string each?

https://github.com/pupbrained/nix-config/blob/2500e475f5314d6c48f8e1b9ef49725f8004ebc2/sys/nix/hardware-configuration.nix#L23

**Comment by pupbrained** [comment]
@lilyinstarlight that fixes the issue. I'm not sure why I even had those all in single strings, honestly.

**Comment by lilyinstarlight** [comment]
> that fixes the issue. I'm not sure why I even had those all in single strings, honestly.

No worries! I'm glad it was that easy to fix and not something wrong here :)

**Pull request merged**
